atlassian/rest_client.py

```diff
@@ -221,17 +265,36 @@ def request(
             json_dump = None if not json else dumps(json)
         self.log_curl_debug(method=method, url=url, headers=headers, data=data if data else json_dump)
         headers = headers or self.default_headers
-        response = self._session.request(
-            method=method,
-            url=url,
-            headers=headers,
-            data=data,
-            json=json,
-            timeout=self.timeout,
-            verify=self.verify_ssl,
-            files=files,
-            proxies=self.proxies,
-        )
+
+        backoff = 1
+        retries = 0
+        responseloop = True
+        while responseloop:
+            response = self._session.request(
+                method=method,
+                url=url,
+                headers=headers,
+                data=data,
+                json=json,
+                timeout=self.timeout,
+                verify=self.verify_ssl,
+                files=files,
+                proxies=self.proxies,
+            )
+            responseloop = False
+            if self.backoff_and_retry:
+                if retries > self.max_backoff_retries:
+                    log.warning("Hit max backoff retry limit of {0}, no more retries.".format(self.max_backoff_retries))
+                    responseloop = False
+                for em in self.retry_error_matches:
+                    if response.status_code == em[0] and response.reason == em[1]:
+                        log.warning('Backing off due to error "{0}: {1}" for {2}s'.format(em[0], em[1], backoff))
+                        time.sleep(backoff)
```

Comment:
If `Retry-After` is send in the Header this should be used, see https://developer.atlassian.com/cloud/jira/platform/rate-limiting/#retry-after-headers.

Reply:
@Spacetown this is a great suggestion. The way it was implemented is based on the same article further down: https://developer.atlassian.com/cloud/jira/platform/rate-limiting/#request-backoff

> Request backoff

> Apps should treat 429 responses as a signal to alleviate pressure on an endpoint. Apps should retry the request only after a delay. Best practice to double the delay after each successive 429 response from a given endpoint. Backoff delays only need to exponentially increase to a maximum value at which point the retries can continue with the fixed maximum delay. You should also apply [jitter](https://en.wikipedia.org/wiki/Jitter)
to the delays to avoid the [thundering herd problem](https://en.wikipedia.org/wiki/Thundering_herd_problem)

100% agree a more complete implementation would include the `Retry-After`. The exponential backoff and retry covers all scenarios whether the header is returned or not so is the minimum viable functionality. At present I don't have capacity to add the `Retry-After` logic and for this version I'd suggest it's unneeded functionality that would delay this PR being merged. This suggestion would be a great GH issue though for future improvement.

Reply:
Jitter done